### PR TITLE
HPCC-28368 Roxie may be blocked by non-SSL probe connections to SSL port

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: roxie/ccd/ccdprotocol.cpp [hunk 1]
```diff
@@ -1523,12 +1533,13 @@ class RoxieSocketWorker : public ProtocolQueryWorker
...
     virtual void init(void *_r) override
     {
-        client.setown(new CSafeSocket((ISocket *) _r));
+        rawClient.setown((ISocket *) _r);
         ProtocolQueryWorker::init(_r);
     }
 
     virtual void threadmain() override
     {
+        client.setown(new CSafeSocket(rawClient.getClear()));
```
Comment (mckellyln) [inline roxie/ccd/ccdprotocol.cpp:1542]: Still do not think this runs the [potentially blocking] secure_accept() in the pooled thread ?
Perhaps:
client.setown(new CSafeSocket(createSecureSocket(rawClient.getClear());
or something like that ?  Handling exceptions etc.

File: roxie/ccd/ccdprotocol.cpp [hunk 2]
```diff
@@ -286,50 +331,14 @@ class ProtocolSocketListener : public ProtocolListener
         while (running)
         {
             Owned<ISocket> client = socket->accept(true);
-            Owned<ISecureSocket> ssock;
             if (client)
             {
-                if (streq(protocol.str(), "ssl"))
+                if (isSSL)
```
Comment (mckellyln) [inline roxie/ccd/ccdprotocol.cpp:336]: I think we can remove this if (isSSL) block

Comment (richardkchapman) [reply roxie/ccd/ccdprotocol.cpp]: Just checking you were awake... or proving that I was not, or something...

Pushed another fix

**Comment by mckellyln** [comment]
Looks good.  Let me test it and then I'll approve.

File: roxie/ccd/ccdprotocol.cpp [hunk 3]
```diff
@@ -1523,12 +1522,13 @@ class RoxieSocketWorker : public ProtocolQueryWorker
...
     virtual void init(void *_r) override
     {
-        client.setown(new CSafeSocket((ISocket *) _r));
+        rawClient.setown((ISocket *) _r);
         ProtocolQueryWorker::init(_r);
     }
 
     virtual void threadmain() override
     {
+        client.setown(new CSafeSocket(listener->createSecureSocket(rawClient.getClear())));
```
Comment (mckellyln) [inline roxie/ccd/ccdprotocol.cpp:1531]: I think createSecureSocket() does not throw, should we check here if client is nullptr and avoid the doMain() call ?

**Review by mckellyln — COMMENTED** [review]
I've tested this with a port checker that holds on to the port - and it works great.
Just one last comment about if there is ever a problem in the secure_accept() code path - do we throw/skip calling doMain() ?

File: roxie/ccd/ccdprotocol.cpp (hunk 3 above)
Comment (ghalliday) [reply roxie/ccd/ccdprotocol.cpp:1531]: Yes.  client will be non null, but the result of createSecureSocket may be null.

Comment (mckellyln) [reply roxie/ccd/ccdprotocol.cpp:1531]: Owned<ISocket> clisock = listener->createSecureSocket(rawClient.getClear());
        if (clisock)
        {
            client.setown(new CSafeSocket(clisock.getClear()));
            doMain("");
        }
        else
            rawClient->shutdownNoThrow();

Comment (mckellyln) [reply roxie/ccd/ccdprotocol.cpp:1531]: Shutdown rawClient to help other side see/feel this ?

**Review by mckellyln — COMMENTED** [review]
Just one last comment about if we should shutdown rawClient if there is a failure ?

**Review by mckellyln — COMMENTED** [review]
I don't think my idea works, sorry.
We should revert this last commit.
Perhaps there is some other way to shutdown socket upon failure, but not sure how just yet.
But would new CSafeSocket(nullptr) always return non-NULL ?

**Review by ghalliday — CHANGES_REQUESTED** [review]

**Review by ghalliday — CHANGES_REQUESTED** [review]
Still another issue

File: roxie/ccd/ccdprotocol.cpp [hunk 4]
```diff
@@ -1523,13 +1522,24 @@ class RoxieSocketWorker : public ProtocolQueryWorker
...
     {
-        client.setown(new CSafeSocket((ISocket *) _r));
+        rawClient.setown((ISocket *) _r);
         ProtocolQueryWorker::init(_r);
     }
 
     virtual void threadmain() override
     {
-        doMain("");
+        ISocket *secure = listener->createSecureSocket(rawClient);
```
[the remark this answers is not in the document]
Comment (mckellyln) [reply roxie/ccd/ccdprotocol.cpp]: Yes, rawClient.getLink() perhaps.

**Review by mckellyln — APPROVED** [review]
Looks good now.

**Review by mckellyln — APPROVED** [review]
Good to add back in compile check.
Approved.

**Review by ghalliday — APPROVED** [review]
please squash

**Pull request merged**